typed-clojure/src/typed_clojure/main.clj

```diff
@@ -0,0 +1,280 @@
+(t/ann negative-failure-f [(t/U t/Str t/Num t/Bool) :-> t/Num])
+(defn negative-failure-f [x]
+  (if (string? x)
+    (count x)
+    (t/ann-form (+ x 1) t/Num))) ; Type error: x could be boolean
+
+;; Example connectives
+;; success
+(t/ann connectives-success-f [(t/U t/Str t/Num) :-> t/Num])
+(defn connectives-success-f [x]
+  (if (not (number? x))
+    (count x)
+    0))
+
+(t/ann connectives-success-g [t/Any :-> t/Num])
+(defn connectives-success-g [x]
+  (if (or (string? x) (number? x))
+    (connectives-success-f x)
+    0))
+
+(t/ann connectives-success-h [(t/U t/Str t/Num t/Bool) :-> t/Num])
+(defn connectives-success-h [x]
+  (if (and (not (boolean? x)) (not (number? x)))
+    (count x)
+    0))
+
+;; failure
+(t/ann connectives-failure-f [(t/U t/Str t/Num) :-> t/Num])
+(defn connectives-failure-f [x]
+  (if (not (number? x))
+    (t/ann-form (+ x 1) t/Num) ; Type error: x could be string
+    0))
+
+(t/ann connectives-failure-g [t/Any :-> t/Num])
+(defn connectives-failure-g [x]
+  (if (or (string? x) (number? x))
+    (t/ann-form (+ x 1) t/Num) ; Type error: x could be string
+    0))
+
+(t/ann connectives-failure-h [(t/U t/Str t/Num t/Bool) :-> t/Num])
+(defn connectives-failure-h [x]
+  (if (and (not (boolean? x)) (not (number? x)))
+    (t/ann-form (+ x 1) t/Num) ; Type error: x is string
+    0))
+
+;; Example nesting_body
+;; success
+(t/ann nesting-body-success-f [(t/U t/Str t/Num t/Bool) :-> t/Num])
+(defn nesting-body-success-f [x]
+  (if (not (string? x))
+    (if (not (boolean? x))
+      (+ x 1)
+      0)
+    0))
+
+;; failure
+(t/ann nesting-body-failure-f [(t/U t/Str t/Num t/Bool) :-> t/Num])
+(defn nesting-body-failure-f [x]
+  (if (or (string? x) (number? x))
+    (if (or (number? x) (boolean? x))
+      (t/ann-form (count x) t/Num) ; Type error: x could be number
+      0)
+    0))
+
+;; Example struct_fields
+;; success
+(t/defalias MyStruct
+  (t/HMap :mandatory {:a (t/U nil Number)}
+          :complete? true))
+(t/ann struct-fields-success-f [MyStruct -> t/Num])
+(defn struct-fields-success-f [x]
+  (if (number? (:a x))
+    (:a x)
+    0))
+
+;; failure
+(t/defalias MyStruct
+  (t/HMap :mandatory {:a (t/U nil Number)}
+          :complete? true))
+(t/ann struct-fields-failure-f [MyStruct -> t/Num])
+(defn struct-fields-failure-f [x]
+  (if (string? (:a x))
+    (t/ann-form (:a x) t/Num) ; Error: can't cast string to Num
+    0))
+
+;; Example tuple_elements
+;; success
+(t/ann tuple-elements-success-f [(t/NonEmptyVec (t/U nil t/Num)) :-> t/Num])
+(defn tuple-elements-success-f [x]
+  (let [first-elem (nth x 0)]
+    (if (number? first-elem)
+      first-elem
+      0)))
+
+;; failure
+(t/ann tuple-elements-failure-f [(t/NonEmptyVec (t/U nil t/Num)) :-> t/Num])
+(defn tuple-elements-failure-f [x]
+  (if (number? (nth x 0))
+    (+ (nth x 0) (nth x 1)) ; Type error: (nth x 1) could be non-number
+    0))
+
+;; Example tuple_length
+;; success
+(t/ann tuple-length-success-f [(t/NonEmptyVec t/Num) :-> t/Num])
```

Comment:
Does Typed Clojure have any fixed-length datatypes?

If not, it can't do this benchmark function ... it's not fair to say it succeeds when the input data structure can only contain Numbers.